### PR TITLE
Add Qwen2-7B model to torchtune.

### PR DESCRIPTION
**Pull request by fyabc**
#### Context

This PR add support to Qwen2-7B, including model components and generation / full-parameter finetune / LoRA finetune recipes.

What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog

- Add support to Qwen2
- Fix code format

#### Test plan

Please make sure to do each of the following if applicable to your PR. (If you're not sure about any one of these just ask and we will happily help.)

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add unit tests for any new functionality
- [x] update docstrings for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
	- [x] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### Implementation Details & Notes

- **transformer decoder**: In Qwen2-0.5B and Qwen2-1.5B, the output projection is replaced with token embedding weights ([`tie_word_embeddings = True`](https://huggingface.co/Qwen/Qwen2-0.5B-Instruct/blob/main/config.json#L21)), so we add a new `torchtune.models.qwen2.transformer.Qwen2TransformerDecoder` module to support this feature (set `output=None`).
- **tokenizer**:  `torchtune.models.qwen2._tokenizer.Qwen2Tokenizer` construct a "fast" tokenizer backed by HuggingFace's *tokenizers* library (follow [official transformers implementation](https://github.com/huggingface/transformers/blob/v4.40.1/src/transformers/models/qwen2/tokenization_qwen2_fast.py)).
  - The official implemtation `torchtune.data._chat_formats.ChatMLFormat` will add an extra `<|IM_END|>` tag at the end of the last assistant message in generation (see [this discussion](https://github.com/pytorch/torchtune/pull/1019/files#r1621540801)), so we override the `tokenize_message()` method to fix this problem (will not add `<|IM_END|>` and EOS token if `messages[-1].role == 'assistant' and not messages[-1].content`)
- **verify the correctness**: We compare the output logits and generated responses between torchtune and official transformer implementation under the same prompts and infer hyperparameters.
- **QLoRA**: Since quantized `torchtune.modules.peft.lora.LoRALinear` does not support bias yet, QLoRA finetune recipes a not added into this PR (will support in future).
- More Qwen2 models (0.5B, 1.5B, etc.) will be added in future.

### CODE REVIEW

**Comment by pbontrager** [comment]
Thank you for this great contribution. Could you provide a bit more information in the PR Description around some of the decisions you made when implementing Qwen2? I'm interested in particular why the TransformerDecoder wasn't sufficient for implementing Qwen. I am also interesting in understanding better how you tested and verified that this implementation is correct. Looking forward to getting this into our library and thank you again.

**Comment by fyabc** [comment]
> Thank you for this great contribution. Could you provide a bit more information in the PR Description around some of the decisions you made when implementing Qwen2? I'm interested in particular why the TransformerDecoder wasn't sufficient for implementing Qwen. I am also interesting in understanding better how you tested and verified that this implementation is correct. Looking forward to getting this into our library and thank you again.

Hi, I have updated some implementation details in the description, please check it again.

**Comment by fyabc** [comment]
> Hi @fyabc, thanks so much for this PR. A couple initial questions:
> 
> 1. The transformer decoder class looks pretty similar to the one we have for [Gemma](https://github.com/pytorch/torchtune/blob/main/torchtune/models/gemma/transformer.py#L18), which also ties the embedding weights. I see there is also an optional post-embedding norm in your implementation, but other than that are these two the same? Basically I am wondering if we should provide a single general `TiedEmbeddingTransformerDecoder` (or something like that) that can be used by both Gemma and Qwen2 models.
> 2. What are the specific differences between Qwen2 RoPE and Llama2/Llama3 RoPE?
> 3. What's the underlying tokenizer used by Qwen2? Is it just GPT-2 BPE tokenization? We currently do not have a dependency in torchtune on the tokenizers library, so adding this may require a bit more discussion.

Thank you for your suggestions!

1. Yes, Qwen2 decoder is the same as Gemma decoder, I think a general `TiedEmbeddingTransformerDecoder` class is a good solution. (or change `TransformerDecoder` directly?)
2. Like [Phi3](https://github.com/pytorch/torchtune/pull/876), Qwen2 RoPE is not numerically equivalent to the Llama2/Llama3 RoPE in bf16 training.
3. Yes, Qwen2 tokenizer is based on GPT-2 BPE tokenization, and I am implementing another version to remove the dependency on tokenizers (derived from [Qwen2Tokenizer](https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2/tokenization_qwen2.py#L83)).

**Comment by ebsmothers** [comment]
> 1. Yes, Qwen2 decoder is the same as Gemma decoder, I think a general `TiedEmbeddingTransformerDecoder` class is a good solution. (or change `TransformerDecoder` directly?)
> 2. Like [Phi3](https://github.com/pytorch/torchtune/pull/876), Qwen2 RoPE is not numerically equivalent to the Llama2/Llama3 RoPE in bf16 training (Qwen2 compute rotary embedding in input dtype, Llama2/3 will convert to fp32).
> 3. Yes, Qwen2 tokenizer is based on GPT-2 BPE tokenization, and I am implementing another version to remove the dependency on tokenizers (derived from [Qwen2Tokenizer](https://github.com/huggingface/transformers/blob/main/src/transformers/models/qwen2/tokenization_qwen2.py?rgh-link-date=2024-07-09T10%3A01%3A49Z#L83)).

Thanks @fyabc for your responses! For (1) let's go with `TiedEmbeddingTransformerDecoder`, no need to change `TransformerDecoder` (since that change would impact a lot more models). If you want to also point Gemma to this new class that'd be great, but if not we can also do it as a follow-up. 

Thanks for clarifying on (2). @joecummings has also been looking at our RoPE implementation so can weigh in with any thoughts he has.

(3) sounds great! Please let me know if you need any pointers from us on adding a new tokenizer; otherwise just ping on here when you are ready for a more detailed review of everything.

**Comment by fyabc** [comment]
@ebsmothers Hi, I have updated this PR based on your suggestions:

1. Add a new `torchtune.modules.transformer.TiedEmbeddingTransformerDecoder` class to handle models with tied embedding weight. (I have not refactored Gemma implementation)
2. Rewrite `Qwen2Tokenizer`, remove the dependency on `tokenizers` library.

File: torchtune/modules/transformer.py
Comment (SalmanMohammadi) [inline torchtune/modules/transformer.py:251]: This PR is awesome!

Small comment here: similar to how `GemmaTransformerDecoder` is a separate component to `TransformerDecoder`, is it worth removing the inheritance here too and just inheriting from `nn.Module`? It should be fine to replicate the code in the constructor and `setup_caches`.

File: torchtune/models/qwen2/_model_builders.py
Comment (SalmanMohammadi) [inline torchtune/models/qwen2/_model_builders.py:23]: similar typing comment here : )

File: torchtune/models/qwen2/_model_builders.py
Comment (SalmanMohammadi) [inline torchtune/models/qwen2/_model_builders.py:1]: Is there generally interest in the 1.5B model too? If so, is it worth adding?

Perhaps as a more general point, I noticed in the Qwen2 technical report that weight tying is only used for the 0.5B and 1.5B models, right? Would users want to configure weight tying in the larger models where weight tying is enabled by default?

File: torchtune/models/qwen2/_tokenizer.py
Comment (SalmanMohammadi) [inline torchtune/models/qwen2/_tokenizer.py:95]: Is it worth thinking about trying to unify the constructor args here to help all our tokenizers look and feel the same?

File: tests/torchtune/models/qwen2/test_qwen2.py
Comment (SalmanMohammadi) [inline tests/torchtune/models/qwen2/test_qwen2.py:1]: I'm sure @ebsmothers will have more thoughts on whether this is necessary here - but I'll share an excellent comment from him about the testing process in torchtune https://github.com/pytorch/torchtune/pull/840#discussion_r1579801172.

File: tests/assets/tiny_bpe_vocab.json
Comment (SalmanMohammadi) [inline tests/assets/tiny_bpe_vocab.json:1]: I've not looked too closely at where this is used, but could we minify this? : )

File: tests/assets/tiny_bpe_tokenizer.json
Comment (SalmanMohammadi) [inline tests/assets/tiny_bpe_tokenizer.json:1]: Similar to my other comment, is it also worth minifying this?

File: torchtune/modules/transformer.py
Comment (joecummings) [reply torchtune/modules/transformer.py:251]: + 1 to this!

File: torchtune/models/qwen2/_tokenizer.py
Comment (joecummings) [inline torchtune/models/qwen2/_tokenizer.py:75]: nit: no longer backed by HuggingFace's tokenizers library

File: torchtune/models/qwen2/_tokenizer.py
Comment (joecummings) [inline torchtune/models/qwen2/_tokenizer.py:81]: Can you add a quick comment here on what "merges.txt" is?

Also, missing a few args in the docstrings.

**Review by joecummings — COMMENTED** [review]
Overall, this is coming together very nicely and I appreciate your hard work to add a great new model to torchtune! Most of the comments I had are nits, but I did have one last thing to say on testing.

The individual unit tests are great for proving correctness of individual models, but a E2E run of this recipe on alpaca and comparing the loss curves between this and HuggingFace is a fantastic sanity check to make sure the torchtune users are getting the expected experience. It's also useful to output both loss curves to a Weights & Biases run so we can visually see that they're similar.

File: torchtune/modules/transformer.py
Comment (fyabc) [reply torchtune/modules/transformer.py:251]: > This PR is awesome!
> 
> Small comment here: similar to how `GemmaTransformerDecoder` is a separate component to `TransformerDecoder`, is it worth removing the inheritance here too and just inheriting from `nn.Module`? It should be fine to replicate the code in the constructor and `setup_caches`.

OK, I will remove the inheritance.

File: tests/assets/tiny_bpe_vocab.json
Comment (fyabc) [reply tests/assets/tiny_bpe_vocab.json:1]: I follow the setting in `tiktoken_small.model` in assets (vocab_size == 2000).
I will reduce it to 500.

File: torchtune/models/qwen2/_model_builders.py
Comment (fyabc) [reply torchtune/models/qwen2/_model_builders.py:1]: Yes, I will add other size of models (0.5B / 1.5B / 72B).
In addition, whether weight tying is enabled must be consistent between model initialization (in `_model_builders.py` here) and weight loading (in `_convert_weights.py`, reading configuration from `config.json` file), so I think we should fix the weight tying configuration in default model providers to avoid users from bugs.
Based on the experience of the open source community, most users have no need to modify weight tying when tuning Qwen models. For expert users, they can directly use `torchtune.models.qwen2.qwen2()` to provide correct hyperparameters.

File: torchtune/models/qwen2/_tokenizer.py
Comment (fyabc) [reply torchtune/models/qwen2/_tokenizer.py:81]: merges.txt contains all BPE merge operations, and this file is required to split a single word into byte-level BPE tokens.
You can find a description at <https://github.com/huggingface/transformers/issues/4777>.

File: torchtune/models/qwen2/_tokenizer.py
Comment (fyabc) [reply torchtune/models/qwen2/_tokenizer.py:95]: Can I set default value of `merges_file` to `None`, and infer it from `vocab_file` (`merges_file = merges_file or os.path.join(os.path.dirname(vocab_file), 'merges.txt')`)?

Some users may change `eos_token` and other special tokens when tuning Qwen2, so I keep these keyword arguments for flexibility.

**Comment by fyabc** [comment]
> Overall, this is coming together very nicely and I appreciate your hard work to add a great new model to torchtune! Most of the comments I had are nits, but I did have one last thing to say on testing.
> 
> The individual unit tests are great for proving correctness of individual models, but a E2E run of this recipe on alpaca and comparing the loss curves between this and HuggingFace is a fantastic sanity check to make sure the torchtune users are getting the expected experience. It's also useful to output both loss curves to a Weights & Biases run so we can visually see that they're similar.

Thank you for you suggestions! I will run all recipes and report the results and comparisons between HuggingFace.

File: tests/assets/tiny_bpe_vocab.json
Comment (SalmanMohammadi) [reply tests/assets/tiny_bpe_vocab.json:1]: Sorry @fyabc, my meaning was to use something like https://codebeautify.org/jsonminifier which just removes any unnecessary characters and formatting. The vocab size itself was probably okay!

File: torchtune/models/qwen2/_tokenizer.py
Comment (SalmanMohammadi) [reply torchtune/models/qwen2/_tokenizer.py:95]: HF will always pull the `merges_file` into the same directory as the vocabulary right? That seems like a sensible suggestion, particularly if you think there are cases when a tokenizer comes with a `merges_file`, but a user wouldn't want to use it. If that's never going to be the case we could maybe just remove it from the `__init__` and hardcode loading it if it exists.

Could we also change `vocab_file` to just `path`, which means we can specify the config for this tokenizer the same as we would any other?

This is a slight tangent, and correct me if I'm wrong here, but the `tokenizer.json` format is for the fast fast BPE format, and from a quick look around, we don't support this yet for Llama models, right? Is it worth opening a broader discussion about generalising this tokenizer?

I think we also have some precedent for generally specifying custom tokens (like `eos_id`) in the codebase. cc @joecummings for thoughts here (and to generally make sure I'm not tripping).

File: torchtune/models/qwen2/_model_builders.py
Comment (SalmanMohammadi) [reply torchtune/models/qwen2/_model_builders.py:1]: Thanks for the insight!

This brings me to a broader point and feel free to ignore me if this sounds crazy. If we don't think there would be interest from the community in the 0.5B/1.5B models, which do use weight tying, would we be able to simplify this PR a bunch by just using the original `TransformerDecoder` for the 7B and 72B models, which don't need weight tying? 

If you think it's worth offering this flexibility I think the added decoder is fine, and it'll be nice to share it with Gemma. (also cc @joecummings @ebsmothers here).

**Comment by SalmanMohammadi** [comment]
I've also added a few comments (but feel free to take these with a pinch of salt since I don't bring the experience that the maintainers do!). 

This PR is great and you've generally fit everything into the codebase well. I've raised some general points about whether it's worth generalising the fast BPE tokenizer that you've implemented, and whether there's interest in the smaller weight-tied models, or a need for flexibility around weight tying in the larger models; I'd love to get some input on these points.

**Review by RdoubleA — COMMENTED** [review]
Sorry I'm late to the party - thank you for all the excellent work here! This was a huge effort to add an important model to the library, and you've integrated it into the codebase well.

One general comment I had on tied embeddings - do you know if users will ever want to use Qwen2 without tied embeddings? I noticed on HF this defaults to False. Asking because assuming it's always tied will simplify a lot of the builder logic and typing. For Gemma for example, we took a stance that it will always be tied, but I don't know enough about Qwen to make that call.

On generalizing the fast BPE tokenizer here - I would vote to keep it in Qwen until another model needs to use the same GPT-2 BPE tokenizer (cc @SalmanMohammadi)

File: torchtune/models/qwen2/_model_builders.py
Comment (joecummings) [reply torchtune/models/qwen2/_model_builders.py:1]: Yeah, we've made it pretty far in this PR so I think best course of action is to land this with the weight tying; however, for future reference @SalmanMohammadi is absolutely correct that we prefer to start by landing *only* what is needed, which - in this case - would just be regular Qwen2 without weight tying.

This code will make it easy to add the small models, which is a huge benefit for our memory constrained users, so I'm excited for the quick follow up with those configs :)

File: torchtune/models/qwen2/_tokenizer.py
Comment (fyabc) [reply torchtune/models/qwen2/_tokenizer.py:82]: sorry, forget to update the docstring... I will fix it later

File: torchtune/models/qwen2/_tokenizer.py
Comment (fyabc) [reply torchtune/models/qwen2/_tokenizer.py:89]: sorry, the doctest is directly copied from other class. I will replace it with correct output.

File: torchtune/models/qwen2/_tokenizer.py
Comment (fyabc) [reply torchtune/models/qwen2/_tokenizer.py:95]: As mentioned in this PR, Qwen2 uses ChatML template, but the exist `torchtune.data._chat_formats.ChatMLFormat` class will add an extra eos token (see notes [here](https://github.com/pytorch/torchtune/pull/1143#issue-2392523051)), so I reimplement the template in Qwen2Tokenizer.
(If `ChatMLFormat` can fix the bug of eos token, I think it is also a good choice to use `ChatMLFormat` directly.)

Comment (RdoubleA) [reply torchtune/models/qwen2/_tokenizer.py:95]: Ah ok, thanks for explaining. Sorry I missed that. I created an issue to track this: #1194. Let's leave the template you have here as is.

File: recipes/configs/qwen2/7B_full.yaml
Comment (SalmanMohammadi) [inline recipes/configs/qwen2/7B_full.yaml:1]: Could you add these to the recipe registry please?

https://github.com/pytorch/torchtune/blob/main/torchtune/_recipe_registry.py

**Comment by fyabc** [comment]
> Hi @fyabc! Thanks for your patience as we reviewed this PR - we're still really excited to get the Qwen2 model into torchtune.
> 
> Would you mind taking another quick look at the comments by @SalmanMohammadi and @RdoubleA ? Let me know how else I can help.

Sorry, I am busy on other work last week and forgot to reply. I will update my code on these comments in few days.

**Comment by joecummings** [comment]
> Sorry, I am busy on other work last week and forgot to reply. I will update my code on these comments in few days.

No worries at all! Please do let us know if there's anyway we can help support you further - we appreciate all the work you've already done here

File: torchtune/models/qwen2/_tokenizer.py
Comment (fyabc) [reply torchtune/models/qwen2/_tokenizer.py]: Already changed to `lru_cache()` and add a configurable cache size argument.

**Comment by fyabc** [comment]
> Thanks so much for the updates @fyabc. This is looking great. As per @joecummings's comment above, would it be possible to verify training works as expected?
> 
> > The individual unit tests are great for proving correctness of individual models, but a E2E run of this recipe on alpaca and comparing the loss curves between this and HuggingFace is a fantastic sanity check to make sure the torchtune users are getting the expected experience. It's also useful to output both loss curves to a Weights & Biases run so we can visually see that they're similar.
> 
> I think once you've had a chance to compare a training run, we should be good to land this (though Joe and Rafi may have some more input). Please let us know if there's more we can do to help - we really appreciate all the work you've put in here.
> 

Thank you for your suggestion! I am running finetune recipes and compare them with the official huggingface finetune script. I will report experiment results to wandb and add the share link to PR description.

File: torchtune/modules/transformer.py
Comment (ebsmothers) [inline torchtune/modules/transformer.py:252]: Now that this is in transformer.py as a core module we need to add a unit test for it. I don't think we need to block this PR on it, so filed #1241 as a follow-up.

File: recipes/configs/qwen2/7B_full.yaml
Comment (ebsmothers) [inline recipes/configs/qwen2/7B_full.yaml:6]: I believe the download commands also need to add the `--ignore-patterns` flag to properly download safetensors files (this is slightly annoying but we need to do things this way for the time being)
```suggestion
#   tune download Qwen/Qwen2-7B-Instruct --output-dir /tmp/Qwen2-7B-Instruct --ignore-patterns ""
```

File: torchtune/models/qwen2/_component_builders.py
Comment (ebsmothers) [inline torchtune/models/qwen2/_component_builders.py:50]: Out of curiosity, is this something we anticipate people to experiment with? I think it's fine to support but (a) it can have implications on checkpointing and/or FSDP wrapping logic, and (b) we are relaxing the contract of the builder function a bit to do it. Not a huge concern, but wanna make sure that it's valuable to add this kind of configurability.

**Review by ebsmothers — APPROVED** [review]
This looks great! I left a few small comments but none of them are blocking. Would like clarification on the point about making `tie_word_embeddings` configurable, but regardless I think we can address anything else in a follow-up. Thank you so much for contributing Qwen2 to our library!

File: torchtune/models/qwen2/_component_builders.py
Comment (ebsmothers) [reply torchtune/models/qwen2/_component_builders.py:50]: Edit: nevermind on this point. I see that you mention it being relevant for 0.5B and 1.5B size models. So I think the approach makes sense

**Pull request merged**
